File: GameCore/src/game/core/listeners/IGameListener.java
package game.core.listeners;

import game.core.Square;

/**
 * Слушатель событий на доске.
 *
 * @author <a href="mailto:vladimir.romanov@gmail.com">Romanov V.Y.</a>
 */
public interface IGameListener {
    /**
     * Единственный экземпляр пустой реализации слушателя доски.
     */
    IGameListener EMPTY = new IGameListener() {
        @Override
        public void mouseDown(Square s, int button) {
        }

        @Override
        public void mouseUp(Square s, int button) {
        }
    };

    /**
     * Нажата кнопка мыши над клеткой <b>s</b> доски.
     *
     * @param s      - клеткой доски.
     * @param button - номер кнопки.
     */
    void mouseDown(Square s, int button);

    /**
     * Отпущена кнопка мыши над клеткой <b>s</b> доски.
     *
     * @param s      - клеткой доски.
     * @param button - номер кнопки.
     */
    void mouseUp(Square s, int button);
}

File: GameCore/src/game/core/listeners/IGameListener.kt
package game.core.listeners

import game.core.Square

/**
 * Слушатель событий на доске.
 *
 * @author [Romanov V.Y.](mailto:vladimir.romanov@gmail.com)
 */
interface IGameListener {
    /**
     * Нажата кнопка мыши над клеткой **s** доски.
     *
     * @param s      - клеткой доски.
     * @param button - номер кнопки.
     */
    fun mouseDown(s: Square, button: Int)

    /**
     * Отпущена кнопка мыши над клеткой **s** доски.
     *
     * @param s      - клеткой доски.
     * @param button - номер кнопки.
     */
    fun mouseUp(s: Square, button: Int)

    companion object {
        /**
         * Единственный экземпляр пустой реализации слушателя доски.
         */
        val EMPTY: IGameListener = object : IGameListener {
            override fun mouseDown(s: Square, button: Int) {}
            override fun mouseUp(s: Square, button: Int) {}
        }
    }
}